client/signup/steps/domains/index.jsx

```diff
@@ -274,6 +293,114 @@ class DomainsStep extends Component {
 		const { lastDomainSearched } = step.domainForm ?? {};
 
 		const { suggestion } = step;
+		const isPurchasingItem = suggestion && Boolean( suggestion.product_slug );
+		const siteUrl =
+			suggestion &&
+			( isPurchasingItem
+				? suggestion.domain_name
+				: suggestion.domain_name.replace( '.wordpress.com', '' ) );
+
+		const domainItem = isPurchasingItem
+			? domainRegistration( {
+					domain: suggestion.domain_name,
+					productSlug: suggestion.product_slug,
+			  } )
+			: undefined;
+
+		if ( hasDomainInCart( this.props.cart, suggestion.domain_name ) ) {
+			this.removeDomain( suggestion ).then( () => {
+				this.updateDomainCart( {
+					suggestion,
+					domainItem,
+					googleAppsCartItem,
+					isPurchasingItem,
+					siteUrl,
+					shouldHideFreePlan,
+					useThemeHeadstartItem,
+					signupDomainOrigin,
+					lastDomainSearched,
+				} );
+			} );
+		} else {
+			this.addDomain( suggestion ).then( () => {
+				this.updateDomainCart( {
+					suggestion,
+					domainItem,
+					googleAppsCartItem,
+					isPurchasingItem,
+					siteUrl,
+					shouldHideFreePlan,
+					useThemeHeadstartItem,
+					signupDomainOrigin,
+					lastDomainSearched,
+				} );
+			} );
+			this.props.setDesignType( this.getDesignType() );
+			// Start the username suggestion process.
+			siteUrl && this.props.fetchUsernameSuggestion( siteUrl.split( '.' )[ 0 ] );
+		}
+	};
+
+	updateDomainCart = ( {
+		suggestion,
+		domainItem,
+		googleAppsCartItem,
+		isPurchasingItem,
+		siteUrl,
+		shouldHideFreePlan,
+		useThemeHeadstartItem,
+		signupDomainOrigin,
+		lastDomainSearched,
+	} ) => {
+		const domainCart = getDomainRegistrations( this.props.cart );
+		this.props.submitSignupStep(
+			Object.assign(
+				{
+					stepName: this.props.stepName,
+					domainItem,
+					googleAppsCartItem,
+					isPurchasingItem,
+					siteUrl,
+					stepSectionName: this.props.stepSectionName,
+					domainCart,
+				},
+				this.getThemeArgs()
+			),
+			Object.assign(
+				{ domainItem, domainCart },
+				this.isDependencyShouldHideFreePlanProvided() ? { shouldHideFreePlan } : {},
+				useThemeHeadstartItem,
+				signupDomainOrigin ? { signupDomainOrigin } : {},
+				suggestion?.domain_name ? { siteUrl: suggestion?.domain_name } : {},
+				lastDomainSearched ? { lastDomainSearched } : {},
+				{ domainCart }
+			)
+		);
+	};
+
+	submitWithDomain = ( { googleAppsCartItem, shouldHideFreePlan = false, signupDomainOrigin } ) => {
+		const { step } = this.props;
+		const { suggestion } = step;
+
+		if (
+			suggestion &&
+			! suggestion.is_free &&
```

Comment:
It looks like this discards anything you've previously selected, we should add free domains to the card even if it's superficial.

Reply:
We're still figuring out what do to for the free domain (p1696345649589759/1696345060.120629-slack-CKZHG0QCR), so I guess we could ignore this for now 🤔